torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1138
+  def broadcast(self, tensors, opts):
+    root_tensor = tensors[opts.rootTensor]
+    root_rank = opts.rootRank
+    if root_rank != self.rank():
+      with torch.no_grad():
+        root_tensor.zero_()
+    xm.all_reduce(xm.REDUCE_SUM, [root_tensor], groups=self._mesh)
+
+    return WorkXla([root_tensor])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2355
+  def reduce_scatter(self, output_tensors, input_tensors_list, opts):
+    for input_tensors, output_tensor in zip(input_tensors_list, output_tensors):
+      # Ensure all inputs have the same shape.
+      last_shape = input_tensors[0].shape
+      for i, t in enumerate(input_tensors[1:]):
+        if last_shape != t.shape:
+          raise ValueError(f"Input {i+1}'s shape is different from input {i}: "
+                           f"{t.shape} vs {last_shape}")
+        last_shape = t.shape
+      input_tensor = torch.cat(input_tensors)
+      reduce_type = self._get_reduce_type(opts.reduceOp)
+      groups = self._mesh
+      shard_count = len(groups[0]) if groups else self.size()
+      result = xm.reduce_scatter(
+          reduce_type,
+          input_tensor,
+          scatter_dim=0,
+          shard_count=shard_count,
+          scale=1,
+          groups=groups)
+
+      output_tensor.set_(result)
+
+    return WorkXla(output_tensors)
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2683
+  def barrier(self, opts):
+    return WorkXla()
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1417
+  # `reduce` is not needed by DeepSpeed for now.
+  def reduce(self, *args):
+    raise NotImplementedError
+
+  def allgather_coalesced(self, *args):
+    raise NotImplementedError
+
+  def allreduce_coalesced(self, *args):
+    raise NotImplementedError
+
+  def alltoall(self, *args):
+    raise NotImplementedError
+
+  def alltoall_base(self, *args):
+    raise NotImplementedError
+
+  def gather(self, *args):
+    raise NotImplementedError
+
+  def scatter(self, *args):
+    raise NotImplementedError
+
+  def send(self, *args):
+    raise NotImplementedError
+
+  def recv(self, *args):
+    raise NotImplementedError
+
+  def recv_anysource(self, *args):
+    raise NotImplementedError
+
+  def monitored_barrier(self, *args):
+    raise NotImplementedError
+
+  def Options(self, *args):
+    raise NotImplementedError
+
+
+class WorkXla(Work):
+
+  def __init__(self, cc_tensors=None):
+    '''
+        Args:
+            cc_tensors (List[torch.Tensor]): List of `torch.Tensor`s that
+            have collective communication ops pending.
```

Comment:
If this is user-facing, we will need indents starting from the 2nd line to make argument docs render correctly.